src/charm.py

```diff
@@ -481,741 +43,243 @@ def __init__(self, *args) -> None:  # type: ignore[no-untyped-def]
-        ingress_name = self._ingress_name(svc_hostname)
-        new_spec = kubernetes.client.V1IngressSpec(rules=[rule])
-
-        # The service hostname might be configured to use TLS.
-        if initial_ingress.spec.tls:
-            new_spec.tls = [
-                kubernetes.client.V1IngressTLS(
-                    hosts=[svc_hostname],
-                    secret_name=initial_ingress.spec.tls[0].secret_name,
-                )
-            ]
-        return kubernetes.client.V1Ingress(
-            api_version="networking.k8s.io/v1",
-            kind="Ingress",
-            metadata=kubernetes.client.V1ObjectMeta(
-                name=ingress_name,
-                annotations=initial_ingress.metadata.annotations,
-                labels={CREATED_BY_LABEL: label},
-            ),
-            spec=new_spec,
-        )
-
-    def _define_ingress(self, body: Any) -> None:
-        """Create or update an ingress in kubernetes.
-
-        Args:
-            body: Ingress resource body to configure.
-        """
-        api = self._networking_v1_api()
-        self._look_up_and_set_ingress_class(api, body)
-        ingress_name = body.metadata.name
-        ingresses = api.list_namespaced_ingress(namespace=self._namespace)
-        if ingress_name in [x.metadata.name for x in ingresses.items]:
-            api.replace_namespaced_ingress(
-                name=ingress_name,
-                namespace=self._namespace,
-                body=body,
-            )
-            LOGGER.info(
-                "Ingress updated in namespace %s with name %s",
-                self._namespace,
-                ingress_name,
+            InvalidIngressError: If both "nginx-route" and "ingress" relations are present
+                or some definition are invalid.
+        """
+        if not self.unit.is_leader():
+            raise InvalidIngressError(
+                "this charm only supports a single unit, "
+                "please remove the additional units "
+                f"using `juju scale-application {self.app.name} 1`"
             )
-        else:
-            api.create_namespaced_ingress(
-                namespace=self._namespace,
-                body=body,
-            )
-            LOGGER.info(
-                "Ingress created in namespace %s with name %s",
-                self._namespace,
-                ingress_name,
+        nginx_route_relation = self.model.get_relation("nginx-route")
+        ingress_relation = self.model.get_relation("ingress")
+        if nginx_route_relation is not None and ingress_relation is not None:
+            raise InvalidIngressError(
+                "nginx-ingress-integrator cannot establish more than one relation at a time"
             )
 
-    def _remove_ingress(self, ingress_name: str) -> None:
-        """Remove ingress resource.
+    def _reconcile(self, definition: IngressDefinition) -> kubernetes.client.V1Ingress:
```

Comment:
This shouldn't return a k8s client object, let's find a way to not need this return value, e.g., retrieving it later

Reply:
Updated, thanks!